frontend/catalyst/jax_primitives.py

```diff
@@ -1320,65 +1363,87 @@ def _while_loop_lowering(
 # for loop
 #
 @for_p.def_abstract_eval
-def _for_loop_abstract_eval(*args, body_jaxpr, **kwargs):
-    return body_jaxpr.out_avals
+def _for_loop_abstract_eval(*args, body_jaxpr, nimplicit, preserve_dimensions, **kwargs):
+    _assert_jaxpr_without_constants(body_jaxpr)
+
+    return infer_output_type_jaxpr(
+        [],
+        body_jaxpr.jaxpr.invars,
+        body_jaxpr.jaxpr.outvars[nimplicit:],
+        expansion_strategy=for_loop_expansion_strategy(preserve_dimensions),
+        num_implicit_inputs=nimplicit,
+    )
 
 
 # pylint: disable=too-many-arguments
 @for_p.def_impl
 def _for_loop_def_impl(
-    ctx, lower_bound, upper_bound, step, *iter_args_plus_consts, body_jaxpr, body_nconsts
+    ctx,
+    lower_bound,
+    upper_bound,
+    step,
+    *iter_args_plus_consts,
+    body_jaxpr,
+    nimplicit=0,
+    body_nconsts,
+    preserve_dimensions,
 ):  # pragma: no cover
     raise NotImplementedError()
 
 
-# pylint: disable=too-many-statements, too-many-arguments
 def _for_loop_lowering(
     jax_ctx: mlir.LoweringRuleContext,
-    lower_bound: ir.Value,
-    upper_bound: ir.Value,
-    step: ir.Value,
     *iter_args_plus_consts: tuple,
     body_jaxpr: core.ClosedJaxpr,
     body_nconsts: int,
     apply_reverse_transform: bool,
+    nimplicit: int,
+    preserve_dimensions,
 ):
     # Separate constants from iteration arguments.
     # The MLIR value provided by JAX for the iteration index is not needed
     # (as it's identical to the lower bound value).
+
     body_consts = iter_args_plus_consts[:body_nconsts]
-    loop_index = iter_args_plus_consts[body_nconsts]
-    loop_args = iter_args_plus_consts[body_nconsts + 1 :]
+    body_implicits = iter_args_plus_consts[body_nconsts : body_nconsts + nimplicit]
+    lower_bound = iter_args_plus_consts[body_nconsts + nimplicit + 0]
+    upper_bound = iter_args_plus_consts[body_nconsts + nimplicit + 1]
+    step = iter_args_plus_consts[body_nconsts + nimplicit + 2]
+    loop_index = iter_args_plus_consts[body_nconsts + nimplicit + 3]
```

Comment:
Is it backward compatible if dynamic type is disabled by the user?

Reply:
Could you please clarify what is dynamic tape? (by default, everything is expected to be backward-compatible with statically-shaped tensors)